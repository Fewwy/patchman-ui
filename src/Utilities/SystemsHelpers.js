import React from 'react';
import searchFilter from '../PresentationalComponents/Filters/SearchFilter';
import staleFilter from '../PresentationalComponents/Filters/SystemStaleFilter';
import systemsUpdatableFilter from '../PresentationalComponents/Filters/SystemsUpdatableFilter';
import { buildFilterChips, templateDateFormat } from './Helpers';
import { intl } from './IntlProvider';
import messages from '../Messages';
import { packageSystemsColumns } from '../SmartComponents/Systems/SystemsListAssets';
import { defaultCompoundSortValues } from './constants';
import { patchSetDetailColumns } from '../SmartComponents/PatchSetDetail/PatchSetDetailAssets';
import { Link } from 'react-router-dom';
import isEmpty from 'lodash/isEmpty';

export const buildFilterConfig = (search, filter, apply, osFilterConfig) => ({
    items: [
        searchFilter(
            apply,
            search,
            intl.formatMessage(messages.labelsFiltersSystemsSearchTitle),
            intl.formatMessage(
                messages.labelsFiltersSystemsSearchPlaceholder
            )
        ),
        staleFilter(apply, filter),
        systemsUpdatableFilter(apply, filter),
        ...osFilterConfig
    ]
});

export const buildTemplateFilterConfig = (search, apply, osFilterConfig) => ({
    items: [
        searchFilter(
            apply,
            search,
            intl.formatMessage(messages.labelsFiltersSystemsSearchTitle),
            intl.formatMessage(
                messages.labelsFiltersSystemsSearchPlaceholder
            )
        ),
        ...osFilterConfig
    ]
});

export const buildActiveFiltersConfig = (filter, search, deleteFilters) => ({
    filters: buildFilterChips(filter, search, intl.formatMessage(messages.labelsFiltersSystemsSearchTitle)),
    onDelete: deleteFilters,
    deleteTitle: intl.formatMessage(messages.labelsFiltersReset)
});

export const systemsColumnsMerger = (defaultColumns, additionalColumns) => {
    let lastSeen = defaultColumns.filter(({ key }) => key === 'updated');
    lastSeen = [{ ...lastSeen[0], key: 'last_upload', sortKey: 'last_upload' }];

    const nameColumn = defaultColumns.filter(({ key }) => key === 'display_name');
    const groupColumn = defaultColumns.filter(({ key }) => key === 'groups');
    const tagsColumn = defaultColumns.filter(({ key }) => key === 'tags');

    groupColumn[0]?.renderFunc ? groupColumn[0].renderFunc = (value) =>
        isEmpty(value?.groups) ? (
            'N/A'
        ) : (
            //change this link to a InsightsLink after router upgrade
            <a href={`./insights/inventory/groups/${value?.groups[0].id}`}>
                {
                    value?.groups[0].name // currently, one group at maximum is supported
                }
            </a>
        ) : [];

    return [...nameColumn, ...groupColumn, ...tagsColumn, ...additionalColumns(), lastSeen[0]];
};

export const templateSystemsColumnsMerger = (defaultColumns) => {
    let lastSeen = defaultColumns.filter(({ key }) => key === 'updated');
    lastSeen = [{ ...lastSeen[0], key: 'last_upload', sortKey: 'last_upload', renderFunc: value => templateDateFormat(value) }];

    let name = defaultColumns.filter(({ key }) => key === 'display_name');
    let tag = defaultColumns.filter(({ key }) => key === 'tags');

    name = [{
        ...name[0],
        renderFunc: (displayName, id) => <Link to={{ pathname: `/systems/${id}` }}>{displayName}</Link>
    }];

    return [...name, ...tag, ...patchSetDetailColumns, lastSeen[0]];
};

export const createSystemsSortBy = (orderBy, orderDirection, hasLastUpload) => {
    orderBy = (orderBy === 'updated' && !hasLastUpload) && 'last_upload' ||
        (orderBy === 'updated' && hasLastUpload) && packageSystemsColumns[0].key || orderBy;

    let sort = `${orderDirection === 'ASC' ? '' : '-'}${orderBy}`;

    //if orderBy is for a compound column reset sort value to relative compound sort value
    Object.keys(defaultCompoundSortValues).forEach(col => {
        if (col === orderBy) {
            sort = defaultCompoundSortValues[col][orderDirection.toLowerCase()];
        }
    });

    return sort;
};
