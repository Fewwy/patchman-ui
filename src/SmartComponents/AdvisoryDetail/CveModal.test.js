import { Provider, useSelector } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import { storeListDefaults } from '../../Utilities/constants';
import { cveRows, readyCveRows } from '../../Utilities/RawDataForTesting';
import { initMocks } from '../../Utilities/unitTestingUtilities.js';
import CvesModal from './CvesModal';
import { createCvesRows } from '../../Utilities/DataMappers';
import { act } from 'react-dom/test-utils';
import { render, screen } from '@testing-library/react';

initMocks();

jest.mock('react-redux', () => ({
    ...jest.requireActual('react-redux'),
    useSelector: jest.fn()
}));

jest.mock('../../Utilities/DataMappers', () => ({
    ...jest.requireActual('../../Utilities/DataMappers'),
    createCvesRows: jest.fn()
}));

jest.mock('../../Utilities/api', () => ({
    ...jest.requireActual('../../Utilities/api'),
    fetchCvesInfo: jest.fn(() => Promise.resolve('success').catch((err) => console.log(err)))
}));

const mockState = { ...storeListDefaults, rows: cveRows };

const initStore = (state) => {
    const customMiddleWare = () => next => action => {
        useSelector.mockImplementation(callback => {
            return callback({ CvesListStore: state });
        });
        next(action);
    };

    const mockStore = configureStore([customMiddleWare]);
    return mockStore({ CvesListStore: state });
};

let wrapper;
let tempWrapper;
let store = initStore(mockState);

beforeEach(() => {
    wrapper = undefined;
    tempWrapper = undefined;
    store.clearActions();
    useSelector.mockImplementation(callback => {
        return callback({ CvesListStore: mockState });
    });
    createCvesRows.mockImplementation(() => readyCveRows);

    wrapper = mount(<Provider store={store}>
        <Router><CvesModal/></Router>
    </Provider>);
    tempWrapper = mount(
        <Provider store={store}>
            <Router><CvesModal /></Router>
        </Provider>
    );
});

afterEach(() => {
    useSelector.mockClear();
    wrapper.unmount();
    tempWrapper.unmount();
});

describe('CveModal.js', () => {
    it('Should match the snapshots', () => {
        const { asFragment } = render(
            <Provider store={store}>
                <Router><CvesModal cveIds={['testCveID']}  /></Router>
            </Provider>
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it('should set rows to undefined to close the modal', () => {
        const { container } = render(
            <Provider store={store}>
                <Router><CvesModal cveIds={['testCveID']}  /></Router>
            </Provider>
        );
        screen.debug(container.querySelector('[data-ouia-component-type="PF4/ModalContent"]'));
        const svgs = screen.getAllByRole('img', { hidden: true });
        expect(svgs[0]).not.toBeNull();
    });

    it('should handle page change', () => {

        const handlePageChange = tempWrapper.find('TableView').props().onSetPage;
        act(() => handlePageChange('', 2));
        act(() => tempWrapper.update());
        expect(tempWrapper.find('TableView').props().store.rows).toEqual(
            [{ cells: [{ title: '[Object] ' }, { title: '[Object]', value: 'Moderate' },
                { title: '7.3' }], id: 'CVE-2021-29931', key: 'CVE-2021-29931' },
            { cells: [{ title: '[Object] ' }, { title: '[Object]', value: 'Moderate' },
                { title: '7.3' }], id: 'CVE-2021-29932', key: 'CVE-2021-29932' }]
        );
        tempWrapper.unmount();
    });

    it('should handle perPage change', () => {

        const handlePerPageChange = tempWrapper.find('TableView').props().onPerPageSelect;
        act(() => handlePerPageChange('', 20));
        act(() => tempWrapper.update());
        expect(tempWrapper.find('TableView').props().store.rows).toEqual(readyCveRows);
        tempWrapper.unmount();
    });

    it('should handle sorting', () => {

        const handleSort = tempWrapper.find('TableView').props().onSort;
        act(() => handleSort('', 0, 'desc'));
        act(() => tempWrapper.update());

        expect(tempWrapper.find('TableView').props().store.rows).
        toEqual(readyCveRows.slice(0, 10).sort(({ id: aId }, { id: bId }) =>  {
            return !aId.localeCompare(bId);
        }));
        tempWrapper.unmount();
    });
});
