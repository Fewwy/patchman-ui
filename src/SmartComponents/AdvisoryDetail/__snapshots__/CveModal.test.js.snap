// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CveModal.js Should match the snapshots 1`] = `<DocumentFragment />`;
